Add ephemeral storage support for localStorage and sessionStorage

mrobinson opened the pull request:
<!-- Add brave-browser issue bellow that this PR will resolve -->
Resolves https://github.com/brave/brave-browser/issues/11619

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

 * Open browser with clean profile, pointed to rewards prod.
 * Enable third-party cookies
 * Enable "Ephemeral Storage" (#brave-ephemeral-storage) in `about:flags`
 * Navigate to https://dev-pages.brave.software/storage/frames.html
 * Ensure that the test on that page works properly.

## Reviewer Checklist:

- [x] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [x] Adequate test coverage exists to prevent regressions 
- [x] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

mrobinson commented:
Apart from review, this is still blocked on the addition of a runtime flag for this feature and running it through the CI.

mrobinson commented:
cc @pes10k

pes10k commented:
Thanks @mrobinson ! Let me know once the flag is in place, and im happy to start using this as a daily build.  Also also, feel free to let me know if you're at the stage where additional tests on https://dev-pages.brave.software/storage/index.html would be helpful

mrobinson commented:
I believe this change is ready for review.

mrobinson commented:
I have pulled out the addition of a runtime flag for all of ephemeral storage to #6615. This PR now depends on that one.

bridiver commented on browser/net/brave_network_delegate_browsertest.cc:
```diff
@@ -3,15 +3,15 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this file,
  * You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+#include "brave/browser/net/brave_network_delegate_browsertest.h"
```
no header file is needed for the tests, everything can be done in the cc file which is the convention

bridiver replied:
also I don't understand the purpose of this change, the indentation change from adding the header file makes it extremely difficult to see what actually changed in the the test

mrobinson replied:
We have added the header so that we can subclass the `BraveNetworkDelegateBrowsertest` fixture. This allows us to avoid a pretty significant amount of code reuse.

bridiver commented on chromium_src/components/content_settings/core/common/cookie_settings_base.cc:
```diff
@@ -119,6 +121,10 @@ bool BraveIsAllowedThirdParty(
       return true;
   }
 
+  if (base::FeatureList::IsEnabled(content_settings::kBraveEphemeralStorage)) {
```
This change still makes no sense to me. If you want to allow all 3rd-party cookies, this is not the way to do it. This is a allow list, not a block list and this is going to break things for cookies.

bridiver replied:
This method has nothing to do with DOM storage, BraveContentSettingsManagerDelegate::AllowStorageAccess controls access to dom storage

mrobinson replied:
I believe that method calls `ShouldStoreState` which eventually calls `ContentSettingsManagerImpl::AllowStorageAccess` which calls this method. This means that the access list here also applies to DOM Storage.

In any case, my thought here is that the flag can be independent of enabling third-party cookies for now. In this setup testing ephemeral storage would require turning on the ephemeral storage flag and also enabling third-party cookies from settings.

mrobinson replied:
This should no longer be a concern because we aren't going to selectively enable storage when the flag is one yet.

bridiver commented on components/ephemeral_storage/browser/ephemeral_storage_tab_helper.cc:
```diff
@@ -0,0 +1,88 @@
+
+#include "brave/components/ephemeral_storage/browser/ephemeral_storage_tab_helper.h"
+
+#include <string>
```
add a space below, system headers should be grouped separately

mrobinson replied:
Done.

bridiver commented on chromium_src/components/content_settings/core/common/features.h:
```diff
@@ -0,0 +1,17 @@
+namespace content_settings {
+
+COMPONENT_EXPORT(CONTENT_SETTINGS_FEATURES)
+extern const base::Feature kBraveEphemeralStorage;
```
this belongs in component/ephemeral_storage, not content_settings

mrobinson replied:
It's here because it is used by `chromium_src/components/content_settings/core/common/cookie_settings_base.cc ` which is compiled into the network service. This means that making it part of `components/ephemeral_storage` isn't possible, or at least I couldn't figure out a way to not create a layering violation. If we can do away with the changes to `cookie_settings_base.cc` this can move to `components/ephemeral_storage` though.

bridiver commented on chromium_src/third_party/blink/renderer/modules/storage/dom_window_storage.cc:
```diff
@@ -8,47 +8,160 @@
-      DOMWindowStorage::From(window).sessionStorage(exception_state);
-  MaybeClearAccessDeniedException(storage, window, &exception_state);
-  return storage;
+const char BraveDOMWindowStorage::kSupplementName[] = "DOMWindowStorage";
```
you can't use the same name as DOMWindowStorage for a different Supplement

mrobinson replied:
The documentation in `supplementable.h` seems to tell a different story:
```
// The Supplement is expected to use the same const char* string instance
// as its key. The Supplementable's SupplementMap will use the address of the
// string as the key and not the characters themselves. Hence, 2 strings with
// the same characters will be treated as 2 different keys.
```
Still, it's probably better to modify the string for clarity's sake. This should be fixed now.

bridiver commented on chromium_src/third_party/blink/renderer/modules/storage/dom_window_storage.cc:
```diff
@@ -8,47 +8,160 @@
+};
+
+const char EphemeralStorageNamespaces::kSupplementName[] =
+    "EphemeralSessionStorageNamespaces";
```
this should match the class name

mrobinson replied:
Fixed.

bridiver commented on components/ephemeral_storage/browser/ephemeral_storage_browsertest.cc:
```diff
@@ -0,0 +1,164 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/net/brave_network_delegate_browsertest.h"
```
you definitely should not be doing this, just add these tests to brave_network_delegate_browsertest and do a scoped enabling of the feature

mrobinson replied:
If you prefer all of the tests for ephemeral storage to be in the `brave_network_delegate_browstertest.cc`, we can do this.

mrobinson replied:
We will still need to subclass `BraveNetworkDelegateBrowserTest` because 
```
[19561:19561:0911/111412.678737:FATAL:feature_list.cc(39)] Check failed: feature_overrides_allowed. FeatureList overrides must happen in the test constructor, before BrowserTestBase::SetUp() has run.
```

mrobinson replied:
In the end, we plan to add a decent number of tests for ephemeral storage, so we've maintained the `EphemeralStorageBrowserTest` fixture in its own file, but are no longer sublcassing `BraveNetworkDelegateTest`.

bridiver commented on components/ephemeral_storage/browser/ephemeral_storage_browsertest.cc:
```diff
@@ -0,0 +1,164 @@
+                                              storage_type));
+  }
+}
+
```
we need tests that show proper separation of storage between domains to verify this functionality

bridiver commented on patches/third_party-blink-renderer-modules-storage-storage_namespace.h.patch:
```diff
@@ -0,0 +1,13 @@
+   // Also forces all owned CachedStorageAreas to be reconnected.
+   void ResetStorageAreaAndNamespaceConnections();
+ 
++  const String& namespace_id() { return namespace_id_; }
```
see https://github.com/brave/brave-browser/wiki/Patching-Chromium#subclass-and-override for patching conventions. This should be `BRAVE_STORAGE_NAMESPACE_H`

bridiver replied:
also this isn't necessary because you can get the namespace from the WebViewClient (which you can get from WebLocalFrameImpl::ViewImpl() -> WebViewImpl::Client)

mrobinson replied:
The tricky thing is getting from the execution context in `DOMWindowStorage` to `LocalFrame`. With OOPI, it isn't guaranteed that the `Page` you are dealing with is backed by a `LocalFrame`. It's possible to use `DeprecatedLocalMainFrame` in this case. This doesn't seem like a good long-term solution though. Do you have an idea for a better way that doesn't run into this issue?

bridiver commented on components/ephemeral_storage/browser/ephemeral_storage_browsertest.cc:
```diff
@@ -0,0 +1,164 @@
+  DISALLOW_COPY_AND_ASSIGN(EphemeralStorageBrowserTest);
+};
+
+IN_PROC_BROWSER_TEST_F(EphemeralStorageBrowserTest,
```
I'm confused, I don't see where you've enabled kBraveEphemeralStorage so isn't this testing the default behavior?

bridiver replied:
these tests (and an existing test (LocalStorageTest) are all failing in ci

mrobinson replied:
The test needs to be updated to work with the runtime feature, which is something I uploaded to #6615. If you don't mind reviewing that, I can work on getting this past the CI today.

bridiver commented on chromium_src/third_party/blink/renderer/modules/storage/dom_window_storage.cc:
```diff
@@ -8,47 +8,160 @@
 // static
-StorageArea* BraveDOMWindowStorage::localStorage(LocalDOMWindow& window,
-                                            ExceptionState& exception_state) {
+StorageArea* BraveDOMWindowStorage::sessionStorage(
```
how do we know whether the feature is enabled here or not?

mrobinson replied:
We don't. We had assumed that if third-party DOM was allowed at all (which is controlled now by `cookie_settings_base.cc`) then the feature was enabled. If this assumption isn't a good one, we will need to plumb through the feature to this place.

mrobinson replied:
I've modified this to use the feature now that it's available here with the most recent changes to #6615.

bridiver commented on chromium_src/third_party/blink/renderer/modules/storage/dom_window_storage.cc:
```diff
@@ -8,47 +8,160 @@
 
-BraveDOMWindowStorage::BraveDOMWindowStorage() {}
-BraveDOMWindowStorage::~BraveDOMWindowStorage() {}
+  String session_storage_id = session_storage_namespace->namespace_id() +
```
I don't understand the point of creating these new namespaces, why are we just using the existing one?

mrobinson replied:
In this case we are creating namespaces for `localStorage` and `sessionStorage` if the `EphemeralStorageNamespaces` hasn't been been added as a supplement to the `Page` yet. We don't have an existing one apart from the namespace used for non-ephemeral-`sessionStorage`, which cannot be used for the ephemeral version. A `Page` may contain both types.

bridiver replied:
`sessionStorage` is ephemeral and the new namespaces you are always (correctly) creating these with `kSessionStorage` just like the original `sessionStorage`

bridiver replied:
basically `sessionStorage` should stay as normal `sessionStorage`, but with the additional caveat that we will clear it on main frame tld change. We only need a new namespace for `localStorage` so it can act like `sessionStorage` in this context

mrobinson replied:
If we clear `sessionStorage` we will also be clearing any `sessionStorage` for the first-party domain. Essentially ephemeral storage can always co-exist in a tab with non-ephemeral storage. We also can't track third-party domains and only clear those, because if you visit

 1. a.com
 2. b.com (a.com in an iframe)
 3. a.com

we want a.com's `sessionStorage` to be preserved between 1 and 3, but to be isolated from a.com's session storage in 2.

bridiver replied:
it should only be cleared for this namespace which is unique per tab

bridiver replied:
and that scenario is incorrect. When this feature is enabled, if the eTLD changes the sessionStorage should be cleared

bridiver replied:
I guess there was a bit of a misunderstanding between myself and @pes10k, so I understand now why you're separating it from the normal sessionStorage

bridiver commented on chromium_src/third_party/blink/renderer/modules/storage/dom_window_storage.cc:
```diff
@@ -8,47 +8,160 @@
+  Document* document = window->GetFrame()->GetDocument();
+  auto storage_area =
+      namespaces->local_storage()->GetCachedArea(document->GetSecurityOrigin());
+  ephemeral_session_storage_ =
```
you are always initializing this with the local storage namespace and you only have one storage area (ephemeral_session_storage_), but session and local storage need to use separate storage areas

bridiver replied:
it really seems like you're overcomplicating things here by instantiating BraveDOMWindowStorage and having a local var. It seems like it would be better to rename `EphemeralStorageNamespaces` -> `EphemeralDOMWindowStorage` and have it own the StorageAreas

mrobinson replied:
I'm not sure I understand your first comment, but I can see if it's possible to simplify this.

mrobinson replied:
The reason why there are two classes here, which is reflected in the original design in blink is that one class is responsible for passing down the `StorageNamespace` from the tab (`Page`) to all of the `DOMWindow`s. This is why `EphemeralStorageNamespaces` is a `Page` supplement and `BraveDOMWindowStorage` is a `DOMWindow` supplement. I believe the reason for this is, like the previous comment, we cannot simply reach into `LocalFrame` here. I'm not sure it is possible to make the `StorageNamespace` per-DOMWindow without:
 1. Breaking an assumption about OOPI
 2. Continuing to store the namespace id as a string as a page `Page` and then duplicating `StorageNamespace` instances per-`DOMWindow`. This doesn't seem like a better solution. Thoughts?

mrobinson replied:
> you are always initializing this with the local storage namespace and you only have one storage area (ephemeral_session_storage_), but session and local storage need to use separate storage areas

I'll fix this. Thanks for pointing this out.

bridiver commented on chromium_src/third_party/blink/renderer/modules/storage/dom_window_storage.cc:
```diff
@@ -8,47 +8,160 @@
+  auto storage_area =
+      namespaces->local_storage()->GetCachedArea(document->GetSecurityOrigin());
+  ephemeral_session_storage_ =
+      StorageArea::Create(document->GetFrame(), std::move(storage_area),
```
I'm confused about how this works. When does the namespace get added to `storage::SessionStorageImpl`? It seems like it should hit this
```  auto found = namespaces_.find(namespace_id);
  if (found == namespaces_.end()) {
    std::move(callback).Run(/*success=*/false);
    return;
  }```

cathiechen replied:
Yes, it seems we need to create `StorageNamespace`s on the browser side. Otherwise, the binds would fail and the data is stored in StoreageAreaMap on the renderer side. That explains why `ClearEphemeralStorage` not effective.
We created a PL to fix this https://github.com/brave/brave-core/pull/6645
PTAL, thanks:)

bridiver commented on components/ephemeral_storage/browser/BUILD.gn:
```diff
@@ -0,0 +1,16 @@
+  ]
+
+  deps = [
+    "//base",
```
This list of deps does not appear to correspond to the headers in ephemeral_storage_tab_helper.cc/ephemeral_storage_tab_helper.h, where did you come up with these?

mrobinson replied:
Fixed. This was a copy-and-paste issue. Thanks for pointing it out.

bridiver commented on components/ephemeral_storage/browser/ephemeral_storage_tab_helper.cc:
```diff
@@ -0,0 +1,88 @@
+          ->GetSessionStorageNamespace(site_instance)
+          ->id() +
+      "ephemeral-session-storage";
+  session_storage_control->DeleteNamespace(session_namespace_id,
```
I think you should be using DOMStorageContext::DeleteSessionStorage/DeleteLocalStorage to avoid using anything in `content/browser` which isn't allowed as a dep here

mrobinson replied:
`DOMStorageContext::DeleteLocalStorage` takes an origin and we need to delete the `LocalStorage` for every third-party origin. I'm not sure it's appropriate here.

mrobinson replied:
In the latest version of the change, I've moved this to `BrowserContext` to avoid the layering violation.

bridiver commented on components/ephemeral_storage/browser/ephemeral_storage_tab_helper.h:
```diff
@@ -0,0 +1,42 @@
+#ifndef BRAVE_COMPONENTS_EPHEMERAL_STORAGE_BROWSER_EPHEMERAL_STORAGE_TAB_HELPER_H_
+#define BRAVE_COMPONENTS_EPHEMERAL_STORAGE_BROWSER_EPHEMERAL_STORAGE_TAB_HELPER_H_
+
+#include "content/public/browser/storage_partition_config.h"
```
why is this here?

mrobinson replied:
Fixed. This was left over from a previous version of the change.

bridiver commented on components/ephemeral_storage/browser/ephemeral_storage_tab_helper.h:
```diff
@@ -0,0 +1,42 @@
+  friend class content::WebContentsUserData<EphemeralStorageTabHelper>;
+
+  WEB_CONTENTS_USER_DATA_KEY_DECL();
+  DISALLOW_COPY_AND_ASSIGN(EphemeralStorageTabHelper);
```
you're missing base/macros.h for this, but `DISALLOW_COPY_AND_ASSIGN` is deprecated

mrobinson replied:
Fixed.

bridiver commented on components/ephemeral_storage/browser/ephemeral_storage_tab_helper.cc:
```diff
@@ -0,0 +1,55 @@
+  if (navigation_handle->IsSameDocument())
+    return;
+
+  ClearEphemeralStorage();
```
This is clearing storage on every load and we only want to clear it when the etld+1 changes

mrobinson replied:
My understanding of the requirements from @pes10k is that a main frame load even to the same etld+1 should clear the storage.

bridiver replied:
the 3p storage is cleared even when `ClearEphemeralStorage` is commented out. If we use the tab helper to clear the storage then it's trivial to keep it for the lifetime of the eTLD which is better than keeping it just for the page load, but the code does not appear to be working as you expect because this method is not needed to achieve that behavior.

bridiver replied:
I'm still debugging, but I believe this is happening because the StorageArea you are creating doesn't have a matching namespace in the browser process. I thought that caused it to crash, but that isn't happening. Instead what appears to be happening is that the StorageArea is recreated on every page load (making this call unnecessary)

mrobinson replied:
This isn't great, since we are relying on a side-effect of our misuse of the internal APIs. I think we are going to fix this so that the browser-side has a a properly matching namespace.

bridiver commented on components/ephemeral_storage/browser/ephemeral_storage_tab_helper.cc:
```diff
@@ -0,0 +1,55 @@
+  }
+
+  WebContents* contents = web_contents();
+  contents->GetBrowserContext()->ClearEphemeralStorageForHost(
```
This also is not correct because it will clear all ephemeral storage for this host and we only want to clear the specific namespace ids for this tab

mrobinson replied:
I was under the impression that there is a one-to-one mapping of `RenderViewHost` to widget.

bridiver commented on chromium_src/third_party/blink/renderer/modules/storage/storage_namespace.h:
```diff
@@ -0,0 +1,14 @@
+#define BRAVE_CHROMIUM_SRC_THIRD_PARTY_BLINK_RENDERER_MODULES_STORAGE_STORAGE_NAMESPACE_H_
+
+#define BRAVE_STORAGE_NAMESPACE_H \
+  const String& namespace_id() { return namespace_id_; }
```
as mentioned in a previous version of this PR this patch is not necessary. You can get the namespace id from the WebViewClient doing something like this
```
auto* web_frame = WebLocalFrameImpl::FromFrame(document->GetFrame())
WebViewImpl* webview = web_frame->ViewImpl();
if (webview && !webview->Client())
  webview->Client()->GetSessionStorageNamespaceId()    
```

mrobinson replied:
Okay. I've pushed a new version of this branch that eliminates the patch.

bridiver commented on patches/third_party-blink-public-common-features.h.patch:
```diff
@@ -0,0 +1,12 @@
+ BLINK_COMMON_EXPORT
+ extern const char kBackForwardCacheABExperimentGroup[];
+ 
++BRAVE_FEATURES_H
```
This patch should not be necessary, you can just define the features in the chromium src override in the same namespace

mrobinson replied:
Okay. The latest version of the branch no longer has this patch. PTAL.

bridiver commented on browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:
```diff
@@ -0,0 +1,156 @@
+  return *local_storage_namespace_map.get();
+}
+
+std::set<ephemeral_storage::EphemeralStorageTabHelper*>& active_tab_helpers() {
```
we don't need an extra list here because we can already access them from BrowserList/TabStrip model from the original code

bridiver commented on browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:
```diff
@@ -0,0 +1,156 @@
+  if (domain == previous_domain)
+    return;
+
+  ClearEphemeralStorageIfNecessary(domain);
```
this should be previous domain, we want to clear it when we navigate away, not when we navigate to the domain

bridiver commented on browser/ephemeral_storage/BUILD.gn:
```diff
@@ -10,4 +10,8 @@ source_set("ephemeral_storage") {
   ]
+
+  if (is_android) {
+    deps += [ "//chrome/browser/ui" ]
```
I'm confused, this is needed with and without android

mrobinson replied:
Okay. I've fixed this.

bridiver replied:
where have you fixed this? I don't see any changes

mrobinson replied:
I have fixed it locally, but I will push a new version of the branch once I've responded to all of @iefremov's comments.

bridiver commented on browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:
```diff
@@ -19,6 +22,11 @@
 #include "net/base/registry_controlled_domains/registry_controlled_domain.h"
 #include "third_party/blink/public/common/features.h"
```
missing a dep for this

mrobinson replied:
:+1:

bridiver reviewed (approved):
approved when deps are fixed

bridiver commented on test/BUILD.gn:
```diff
@@ -505,6 +505,7 @@ if (!is_android) {
       "//brave/browser/brave_stats/brave_stats_updater_browsertest.cc",
       "//brave/browser/browsing_data/brave_clear_browsing_data_browsertest.cc",
       "//brave/browser/devtools/brave_devtools_ui_bindings_browsertest.cc",
+      "//brave/browser/ephemeral_storage/ephemeral_storage_browsertest.cc",
```
almost missed this. Adding new files to the main test build target should be avoided. Please put this file in it's own source set in brave/browser/ephemeral_storage/BUILD.gn and add a dep here

bridiver replied:
and make sure to include all necessary deps from the headers here https://github.com/brave/brave-core/pull/6560/files#diff-ad1b5bc543791a83f0e762f32a01e88544b826f9ecdad47ce2b8a49461cf7a7bR8

mrobinson replied:
Okay. I think this should be fixed now.

iefremov commented on chromium_src/content/public/browser/browser_context.h:
```diff
@@ -0,0 +1,30 @@
+CreateSessionStorageNamespace(content::StoragePartition* partition,
+                              const std::string& namespace_id);
+
+CONTENT_EXPORT std::string GetSessionStorageNamespaceId(WebContents*);
```
I probably missed something, but why do we keep these functions in the override and not in a separate file?

bridiver replied:
because the implementation uses `content/browser` so the methods have to be defined inside `content/public/browser` so we can call into them

iefremov commented on browser/ephemeral_storage/BUILD.gn:
```diff
@@ -0,0 +1,17 @@
+source_set("ephemeral_storage") {
+  check_includes = false
```
curious, does it fail `gn check`?

mrobinson replied:
`npm run gn_check` does not give any errors.

iefremov replied:
yeah then why add `check_includes = false`?

mrobinson replied:
Good question. Looks like there is a circular dependency which is used to include `profile.h` in the tab helper.

iefremov replied:
we have a master of Deps @Gyuyoung who could help :) 
in any case, we should only add `check_includes = false` together with a link to a github issue

bridiver replied:
@iefremov this cannot be resolved until the `//chrome/browser` dep issue has been resolved and @Gyuyoung is still working on that

iefremov replied:
yep, so I just want to have an issue number to be mentioned here

mrobinson replied:
Is this an example of @Gyuyoung fixing a similar issue? https://github.com/brave/brave-core/pull/6712/files If so, I can follow the same pattern here.

mrobinson replied:
The addition of `PerTLDEphemeralStorage` in the tab helper makes this dependency unnecessary now, so I should be able to remove the `check_includes = false` line.

iefremov commented on chromium_src/third_party/blink/renderer/modules/storage/dom_window_storage.cc:
```diff
@@ -31,25 +33,213 @@ void MaybeClearAccessDeniedException(StorageArea* storage,
+
+  MaybeClearAccessDeniedException(storage, *window, &exception_state);
+  if (base::FeatureList::IsEnabled(features::kBraveEphemeralStorage)) {
+    storage = ephemeralSessionStorage(storage);
```
why not just unconditionally `return ephemeralSessionStorage(storage)` ? We have the check inside `ephemeralSessionStorage()`

mrobinson replied:
Okay. I'll fix this.

iefremov commented on chromium_src/third_party/blink/renderer/modules/storage/dom_window_storage.cc:
```diff
@@ -31,25 +33,213 @@ void MaybeClearAccessDeniedException(StorageArea* storage,
+  Document* document = window->GetFrame()->GetDocument();
+  ephemeral_local_storage_ =
+      StorageArea::Create(document->GetFrame(), std::move(storage_area),
+                          StorageArea::StorageType::kSessionStorage);
```
Some explanation of why it is session here would indeed be helpful for future readers

mrobinson replied:
:+1:

iefremov commented on chromium_src/third_party/blink/renderer/modules/storage/dom_window_storage.cc:
```diff
@@ -31,25 +33,213 @@ void MaybeClearAccessDeniedException(StorageArea* storage,
+
+  auto* security_origin =
+      page->MainFrame()->GetSecurityContext()->GetSecurityOrigin();
+  String domain = security_origin->RegistrableDomain();
```
just curious whether the protocol should participate in the id? I gather we should have different storages for http and https, does it work like this?

mrobinson replied:
It isn't working like that right now and, in fact, subdomains are also lost here. @bridiver, can you confirm is that's what you expected to happen here?

iefremov commented on chromium_src/third_party/blink/renderer/modules/storage/brave_dom_window_storage.h:
```diff
@@ -6,20 +6,36 @@
 
+  StorageArea* sessionStorage(ExceptionState&);
+  StorageArea* localStorage(ExceptionState&);
+  StorageArea* ephemeralSessionStorage(StorageArea*);
```
probably these two could be private

mrobinson replied:
:+1:

iefremov commented on browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:
```diff
@@ -0,0 +1,183 @@
+using content::NavigationHandle;
+using content::WebContents;
+
+namespace {
```
nit: typically anonymous namespace sits inside a normal namespace (i.e. `ephemeral_storage`)

mrobinson replied:
:+1:

iefremov commented on browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:
```diff
@@ -0,0 +1,183 @@
+}
+
+void EphemeralStorageTabHelper::ClearEphemeralStorageIfNecessary() {
+  if (!base::FeatureList::IsEnabled(blink::features::kBraveEphemeralStorage)) {
```
maybe we can just skip creation of the helper at all?

mrobinson replied:
At some point, this will probably be a feature that can be toggled on and of at runtime. Does it make sense to structure the code as if that were the case?

iefremov replied:
probably yes, but this could be tricky, for instance disabling the feature at runtime will block `ClearEphemeralStorageIfNecessary()`, so existed storage would remain dirty. So I'd rather keep things simple, but I'm not a feature owner

mrobinson replied:
Okay. I'll go with the approach you suggest.

iefremov commented on browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:
```diff
@@ -0,0 +1,183 @@
+      URLToStorageDomain(web_contents()->GetLastCommittedURL());
+  if (!IsAnotherTabOpenWithStorageDomain(previous_domain)) {
+    local_storage_namespace_map().erase(previous_domain +
+                                        "/ephemeral-local-storage");
```
can we please move this `domain + "/ephemeral-local-storage"` to a helper function

mrobinson replied:
:+1:

iefremov commented on browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:
```diff
@@ -0,0 +1,183 @@
+  // GetDomainAndRegistry might return an empty string if this host is an IP
+  // address or a file URL.
+  if (domain.empty())
+    domain = url::Origin::Create(url.GetOrigin()).Serialize();
```
wondering whether it is fully aligned with `security_origin->ToUrlOrigin()` in the renderer. Not a blocker of course, but maybe we could add a test for this

mrobinson replied:
I can modify the blink side to avoid calling `SecurityOrigin::ToUrlOrigin()` and to get the `base::Origin` directly from the Document URL. Hopefully that should take care of the possibility of different serialization.

iefremov replied:
I'm not sure whether this is a proper way, since storages are definitely something security-sensitive and AFAIK we should only use SecurityOrigin for sensitive stuff

mrobinson replied:
The documentation for url::Origin says it should be suitable for security-relevant decisions:

```c++
// TL;DR: If you need to make a security-relevant decision, use 'url::Origin'.
// If you only need to extract the bits of a URL which are relevant for a
// network connection, use 'url::SchemeHostPort'.
```

That said, this shouldn't override any existing origin-based security. Even if one.a.com and one.b.com have the same ephemeral storage namespace, they won't be able to access each other's DOM storage. The namespace takes care internally of enforcing normal web security protocols.

iefremov replied:
sorry, this is a wrong place for my comment - I was actually concerned with using `window->Url()` in `EphemeralStorageNamespaces::From`. I'm still not sure what is the proper way to do what we want, but I can't provide any evidence that the current way is not secure :)

iefremov commented on browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:
```diff
@@ -0,0 +1,183 @@
+  ClearEphemeralStorageIfNecessary();
+
+  auto* browser_context = web_contents()->GetBrowserContext();
+  auto site_instance = content::SiteInstance::CreateForURL(
```
this is not necessary if we wanted just `BrowserContext::GetDefaultStoragePartition`

mrobinson replied:
Hrm. This properly calls `GetStoragePartitionConfigForSite`. Perhaps that only matters for plugins though.

iefremov commented on browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:
```diff
@@ -0,0 +1,183 @@
+      content::GetSessionStorageNamespaceId(web_contents()) +
+      "/ephemeral-session-storage";
+  it = session_storage_namespace_map().find(session_partition_id);
+  if (it == session_storage_namespace_map().end()) {
```
possible to save a couple of lines here and there by checking `session_storage_namespace_map().count > 0`

mrobinson replied:
This should be completely gone in the new branch that I will upload soon.

iefremov commented on browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:
```diff
@@ -0,0 +1,183 @@
+}
+
+bool EphemeralStorageTabHelper::IsAnotherTabOpenWithStorageDomain(
+    const std::string& storage_domain) {
```
`IsAnotherTabOpenWithStorageDomain` probably no need to pass `storage_domain` if this is always `URLToStorageDomain(web_contents()->GetLastCommittedURL())`

mrobinson replied:
This is to avoid converting the previously committed URL to a storage domain two times.

iefremov commented on chromium_src/content/browser/browser_context.cc:
```diff
@@ -0,0 +1,41 @@
+          partition->GetDOMStorageContext());
+
+  return content::SessionStorageNamespaceImpl::Create(context_wrapper,
+                                                      namespace_id);
```
There is an interesting quirk, saying that `namespace_id` length is basically hardcoded [1]. Wondering if this could potentially hit us? If we are fine it's worth mentioning in a comment

[1] https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/common/dom_storage/session_storage_namespace_id.cc;l=17?q=session_storage_namespace_id

mrobinson replied:
Oh, nice find! I will modify the code to use and MD5 hash and pad the hash in a way that it cannot collide with GUID generation.

iefremov commented on browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:
```diff
@@ -0,0 +1,183 @@
+  }
+
+  std::string session_partition_id =
+      content::GetSessionStorageNamespaceId(web_contents()) +
```
this creates the namespace anyway, so we can just get the namespace and DCHECK

iefremov replied:
oops sorry, this is not quite right :D

iefremov commented on chromium_src/content/browser/browser_context.cc:
```diff
@@ -0,0 +1,41 @@
+                                                      namespace_id);
+}
+
+std::string GetSessionStorageNamespaceId(WebContents* web_contents) {
```
this function can be moved out here if we get namespace via web_contents, navigation controller and GetSessionStorageNamespaceMap

mrobinson replied:
The issue, I think, is that the delegate definition isn't part of the public API. It's just a forward-declaration.

iefremov replied:
in theory we can do `web_contents()->GetController().GetSessionStorageNamespaceMap()` and pick it from this by site instance ID. not sure if this is more elegant, but at least achievable via `content/public`

mrobinson replied:
Hrm. It looks like `NavigationControllerImpl::GetSessionStorageNamespace` will either get or create the default namespace. Creating the default namespace is something that can only happen with private API. Wouldn't duplicating this code mean there is a possibility that this code could be executed before the default namespace is created and be able to find the id?

iefremov replied:
not sure, though it seems safer to keep as is, given that the file will stay with us anyways

iefremov commented on browser/ephemeral_storage/ephemeral_storage_browsertest.cc:
```diff
@@ -0,0 +1,302 @@
+
+  ValuesFromFrames GetValuesFromFrames(WebContents* web_contents) {
+    RenderFrameHost* main_frame = web_contents->GetMainFrame();
+    return ValuesFromFrames{
```
no need to specify `ValuesFromFrames`

mrobinson replied:
:+1:

iefremov commented on browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:
```diff
@@ -0,0 +1,183 @@
+  return *session_storage_namespace_map.get();
+}
+
+content::SessionStorageNamespaceMap& local_storage_namespace_map() {
```
I'm slightly concerned that this is global for all profiles. It looks like different helpers from different profiles (including incognito and TOR) will use the same namespace for a given domain. Shouldn't we keep namespaces in a keyed service or something like that?

mrobinson replied:
Great point! I'll take a look at this tomorrow.

mrobinson replied:
Okay. The new version of this branch should now key this information with the BrowserContext as well.

iefremov commented on browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:
```diff
@@ -0,0 +1,183 @@
+    }
+  }
+#else
+  for (Browser* browser : *BrowserList::GetInstance()) {
```
on shutdown we will do this for each destroyed web_contents, thus it looks like O(N^2) ops on UI thread. I'd suggest to reconsider having a data structure for this

mrobinson replied:
Hrm. That's an interesting point and I think it would also be true even if we used another data structure here. It seems a solution could be to special-case shutting down.

iefremov replied:
I'd say special-case window closing

mrobinson replied:
I have spent a bit of time thinking about this and I think that your initial idea of creating a new data structure makes a lot of sense. I have gone ahead and done that. The new design, which I will upload soon does this:
  * Adds a new reference-counted data structure `PerTLDEphemeralStorage` which manages all per-TLD ephemeral storage (currently) only localStorage.
  * Each `EphemeralStorageTabHelper` keeps a reference to this data structure which is fetched from a map of `WeakPtr<PerTLDEphemeralStorage>` when the domain changes.
  * When the last tab helper using this per-TLD storage is freed or switches domains, reference counting ensures that the storage is cleaned up.

I think this new design has a bunch of nice benefits:
  * It's less code.
  * It removes a dependency on the UI bits of the browser that were causing a dependency cycle. We can turn on `check_includes` again without any issues.
  * It's more efficient, because we don't have to iterate through tabs or look into a map. The reference counting code is highly-optimized.
   * We only keep one map instead of two. Session storage is now fully managed by the tab helper.

iefremov commented on browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:
```diff
@@ -0,0 +1,183 @@
+  if (it == session_storage_namespace_map().end()) {
+    auto session_storage_namespace =
+        content::CreateSessionStorageNamespace(partition, session_partition_id);
+    session_storage_namespace_map()[session_partition_id] =
```
I'm not a C++ guru but this probably should be `emplace`

mrobinson replied:
:+1:

iefremov commented on chromium_src/third_party/blink/renderer/modules/storage/dom_window_storage.cc:
```diff
@@ -31,25 +33,213 @@ void MaybeClearAccessDeniedException(StorageArea* storage,
+
+  Document* document = window->GetFrame()->GetDocument();
+  ephemeral_local_storage_ =
+      StorageArea::Create(document->GetFrame(), std::move(storage_area),
```
just `window->GetFrame()`

mrobinson replied:
:+1:

pes10k replied:
Just a note here, since it bit the farbling work recently; will this work in a disconnected frame?  do you need to do the same thing as here? https://github.com/brave/brave-core/pull/6941/files

pes10k replied:
maybe of interest to @iefremov ?

iefremov replied:
Yeah @pes10k I think we are protected from this issue, becasue at first we get "normal" storage from the parent method (e.g. `DOMWindowStorage::sessionStorage()` and it explicitly checks for `!window->GetFrame()`

mrobinson commented:
Okay. I've pushed a new commit that addresses review comments and also rebased the entire branch. I think there is really only one question outstanding and it is: https://github.com/brave/brave-core/pull/6560#discussion_r512105647

It's just waiting on @bridiver for feedback on what the behavior was expected with that code. Thanks for the review comments and your patience with this change.

iefremov commented on browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:
```diff
@@ -145,7 +144,10 @@ bool EphemeralStorageTabHelper::IsAnotherTabOpenWithStorageDomain(
 #else
   for (Browser* browser : *BrowserList::GetInstance()) {
-    if (browser->profile() == web_contents()->GetBrowserContext()) {
+    // We need to use reinterpret_cast here, because BrowserContext is a subclass of Profile,
```
hmmm, it seems to me that it is actually `Profile` subclasses `BrowserContext`, so we can just write `BrowserContext* context = browser->profile()` without a cast

mrobinson replied:
I think the issue here was that including the `profile.h` header introduces a circular dependency. This code should be gone in the latest revision though.

iefremov commented on chromium_src/third_party/blink/renderer/modules/storage/dom_window_storage.cc:
```diff
@@ -31,25 +42,209 @@ void MaybeClearAccessDeniedException(StorageArea* storage,
+  // localStorage area, we use the sessionStorage infrastructure.
+  Document* document = window->GetFrame()->GetDocument();
+  ephemeral_local_storage_ =
+      StorageArea::Create(document->GetFrame(), std::move(storage_area),
```
just `window->GetFrame()`

mrobinson replied:
:+1:

iefremov commented on chromium_src/third_party/blink/renderer/modules/storage/dom_window_storage.cc:
```diff
@@ -31,25 +42,209 @@ void MaybeClearAccessDeniedException(StorageArea* storage,
+
+StorageArea* BraveDOMWindowStorage::ephemeralLocalStorage(
+    StorageArea* non_ephemeral_storage) {
+  if (!base::FeatureList::IsEnabled(features::kBraveEphemeralStorage))
+    return non_ephemeral_storage;
```
I'm thinking that maybe we can move these checks to the upper level? Or somehow change the function name, because now it is actually not `ephemeralLocalStorage`, but "ephemeral local storage if things are good, otherwise normal local storage"

mrobinson replied:
:+1:

iefremov commented on chromium_src/third_party/blink/renderer/modules/storage/dom_window_storage.cc:
```diff
@@ -87,24 +96,26 @@ EphemeralStorageNamespaces* EphemeralStorageNamespaces::From(
+  String session_storage_id = StringToSessionStorageId(
       String::FromUTF8(webview->Client()->GetSessionStorageNamespaceId()) +
-      String("/ephemeral-session-storage");
+      "/ephemeral-session-storage");
```
I'd really like to move this concatenation `+ /ephemeral-session-storage` into `StringToSessionStorageId()`

mrobinson replied:
:+1:

iefremov commented on browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:
```diff
@@ -0,0 +1,139 @@
+// hash and then use that make up our own GUID-like string. Because of the way
+// we are constructing the string we should never collide with a real GUID and
+// we only need to worry about hash collisions, which are unlikely.
+std::string StringToSessionStorageId(const std::string& string) {
```
this function can append `"/ephemeral-local-storage"` to the string

mrobinson replied:
One caller needs the string "/ephemeral-local-storage" and the other needs "/ephemeral-session-storage". What do you think about adding an additional parameter which is the suffix to be added to the string?

mrobinson replied:
I've gone ahead and done this in both `dom_window_storage.cc` and `ephemeral_storage_tab_helper.cc`.

iefremov commented on browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:
```diff
@@ -0,0 +1,139 @@
+  } else {
+    std::string local_partition_id = new_domain + "/ephemeral-local-storage";
+    auto local_storage_namespace = content::CreateSessionStorageNamespace(
+        partition, StringToSessionStorageId(local_partition_id));
```
I'm probably missing something, but if `CreateSessionStorageNamespace` returns same namespace for a given partition (which maps 1-1 to a browser context) and a given id, then we don't need a global map? Since storage partition already maintains one

mrobinson replied:
I believe that `CreateSessionStorageNamespace` returns a reference. If all the references disappear then, we the namespace is deleted. We need the map to maintain at least one reference, keeping the namespace alive.

iefremov replied:
I'm honestly having hard times understanding this, will definitely need some help :) 
How does map helps us to keep a namespace alive if (1) it only holds weak pointers (2) last `per_tld_ephemeral_storage_` will erase the corresponding entry in the map?
So anyway, closing (or navigating away) all tabs with a given TLD cleans up the namespace

mrobinson replied:
The maps holds weak pointers, so the map alone won't keep the PerTLDEphemeralStorage (and the namespace) alive on its own. The TabHelper does keep a real reference to the PerTLDEphemeralStorage though. As long as one TabHelper has a reference, the PerTLDEphemeralStorage will stay alive. When the last TabHelper drops the reference (either by being destroyed or by switching domains during navigation), the PerTLDEphemeralStorage will be deleted by the reference counting code and the destructor removes it from the map of weak pointers.

iefremov commented on browser/ephemeral_storage/ephemeral_storage_browsertest.cc:
```diff
@@ -0,0 +1,302 @@
+}
+
+IN_PROC_BROWSER_TEST_F(EphemeralStorageBrowserTest,
+                       ClosingTabClearsEphemeralStorage) {
```
maybe also check that reload doesn't drop the storage?

iefremov replied:
and also check that storage is different in another profile (should be relatively easy to spinoff an incognito window and do same checks)

iefremov replied:
(both tests can be a follow-up of course)

mrobinson replied:
Okay. This all makes sense and I'm happy to add these tests in a quick followup, as you suggest.

mrobinson commented:
Thanks for the comments. I've pushed a new commit to this branch with fixes.

mrobinson commented:
I've pushed a new branch completely removing `PerTLDEphemeralStorage`. After chatting with @iefremov we agreed that it isn't completely necessary for this change and we can rely completely on the reference-counting of the storage namespaces.

iefremov commented on browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:
```diff
@@ -0,0 +1,110 @@
+                                     const std::string& suffix) {
+  std::string hash = base::MD5String(string) + "____";
+  DCHECK_EQ(hash.size(), 36u);
+  return hash + suffix;
```
hmm wait, I gather we should rather do `md5(hash + suffix)`

mrobinson replied:
Oof. You are right. I will fix this.

iefremov commented on browser/ephemeral_storage/ephemeral_storage_tab_helper.h:
```diff
@@ -0,0 +1,46 @@
+#include <utility>
+
+#include "base/memory/ref_counted.h"
+#include "base/memory/weak_ptr.h"
```
not used now

mrobinson replied:
:+1:

iefremov commented on chromium_src/third_party/blink/renderer/modules/storage/dom_window_storage.cc:
```diff
@@ -31,25 +43,206 @@ void MaybeClearAccessDeniedException(StorageArea* storage,
+    return nullptr;
+  String session_storage_id = StringToSessionStorageId(
+      String::FromUTF8(webview->Client()->GetSessionStorageNamespaceId()),
+      "/ephemeral-session-storage");
```
bad spacing

mrobinson replied:
Do you mean vertical spacing? This particular formatting is the result of running "git cl format."

iefremov replied:
sorry, everything's fine here

iefremov commented:
So far found one minor (probably known) trouble:
1) load https://dev-pages.brave.software/storage/frames.html
2) restart the browser, the page should be restored
3) do some input via the form, see all 3 values are set

4a) reload
5a) the value in the foreign iframe disappeared.

4b) open the same url in another tab
5b) there is no value in the foreign iframe

iefremov commented:
Also, would be super helpful to add some top-level comments for key classes (e.g. the TabHelper, `EphemeralStorageNamespaces`) saying what are they responsible for and how do we connect a namespace in the browser process with renderers.

mrobinson commented:
> So far found one minor (probably known) trouble:
>

It seems this is a regression from removing `PerTLDEphemeralStorage`.

iefremov reviewed (approved):
LGTM (only needs fix for https://github.com/brave/brave-core/pull/6560#discussion_r513717640)

mrobinson commented:
> > It seems this is a regression from removing PerTLDEphemeralStorage.
> 
> nope, actually reproduced with `PerTLDEphemeralStorage` as well.
> Edit: I think it could be related to not having a "previous" URL after session restore

The issue here was that there was a typo in the code generating the hash. I have fixed this now and all the CI are passing. I think this can be landed now?

iefremov commented:
> The issue here was that there was a typo in the code generating the hash. I have fixed this now and all the CI are passing. I think this can be landed now?

Sure thing, please go ahead!
